### PR TITLE
Do not display an error message if the default CT log list is not found

### PR DESCRIPTION
**Pull request by RJPercival**
This saves users getting concerned if they see "Failed to load default Certificate Transparency log list" when using s_client. An error will still be displayed if they explicitly provide a CT log list using --ctlogfile and loading fails.

### CODE REVIEW

File: apps/apps.c [hunk 1]
```diff
@@ -238,11 +238,13 @@ int ctx_set_verify_locations(SSL_CTX *ctx, const char *CAfile,
...
-            BIO_puts(bio_err, "Failed to load default Certificate Transparency "
-                     "log list\n");
-        }
-        return 1; /* Do not treat failure to load the default as an error */
+        /*
+         * Try and load the log list from the default location.
+         * We don't treat it as an error if this fails, since the user didn't
+         * explicitly request a log list be loaded and might not need it either.
+         */
+        SSL_CTX_set_default_ctlog_list_file(ctx);
```
Comment (ekasper) [inline apps/apps.c]: Hm, new boolean functions should generally be __owur (stands for OpenSSL warn unused result, see the header for examples). It's better to make it an error in general, and swallow the error here. Alternatively, you could just downgrade the message to "Info: no default Certificate Transparency log list available."

Also - not for the scope of this PR, but out of curiosity, when does this return '<0'? I think (though that's my personal preference) that we should keep new APIs 1/0 whenever possible.

**Comment by richsalz** [comment]
yes, make the owur change and i will also plus-one

File: apps/apps.c (hunk 1 above)
Comment (RJPercival) [inline apps/apps.c]: I'll create a new PR that adds __owur to all of the CT-related boolean functions.

Failure to load a log is reported as an error (in crypto/ct/ct_log.c), but SSL_CTX_set_default_ctlog_list_file() swallows those errors, which is consistent with the behaviour of SSL_CTX_set_default_verify_dir() and SSL_CTX_set_default_verify_file(). I'd be happy to stop it swallowing errors and suppress them here instead though. It arguably makes most sense to suppress them in s_client, depending on whether -requestct or -requirect were used.

I've checked and this appears to only ever return 1 or 0. While I'm adding __owur, I'll check for anywhere the boolean functions are called and swap "x <= 0" for "!x".

**Comment by RJPercival** [comment]
PR #791 contains the __owur change.

**Comment by richsalz** [comment]
+1

File: apps/apps.c [hunk 2]
```diff
@@ -238,11 +238,7 @@ int ctx_set_verify_locations(SSL_CTX *ctx, const char *CAfile,
 int ctx_set_ctlog_list_file(SSL_CTX *ctx, const char *path)
 {
     if (path == NULL) {
-        if (SSL_CTX_set_default_ctlog_list_file(ctx) <= 0) {
-            BIO_puts(bio_err, "Failed to load default Certificate Transparency "
-                     "log list\n");
-        }
-        return 1; /* Do not treat failure to load the default as an error */
+        return SSL_CTX_set_default_ctlog_list_file(ctx);
```
Comment (kroeckx) [inline apps/apps.c]: Please document this function, including what the return value means.

**Comment by kroeckx** [comment]
So I've been reading the existing documentation in doc/ssl/SSL_CTX_set_ctlog_list_file.pod, and I have a few comments.

It documents as SSL_CTX_set_default_ctlog_list_file() never returning errors.  But it does, or you wouldn't be changing this in the first place.

I'm also not very happy with skipping on error, I would rather that it gave an error.  If it's not working you now have no idea why.

I'm also not sure why we have a default function and why we can't use the normal function with a default path instead.

**Comment by RJPercival** [comment]
Where the documentation says it "never reports errors", that refers to the fact that `SSL_CTX_set_default_ctlog_list_file` calls `ERR_clear_error` and so error reporting is suppressed. The return value could still be used to check for success though. The documentation does say this, but it could certainly be more explicit about what not reporting errors means. As it happens though, this PR removes that error suppression (so I ought to update the doc). It introduces new error suppression, but only in s_client in the specific case that the default log list fails to load and CT support has not explicitly been requested. It seems it would only confuse users to display errors about something failing that they haven't asked for in the first place.

These functions are modelled after the existing certificate loading functions, which work in this way. I'm happy to change it though, if you have concrete recommendations. The current set of default file paths live in crypto/include/internal/cryptlib.h, which aren't accessible from apps. I could put the default path in include/openssl/ct.h, if that's acceptable, and then remove SSL_CTX_set_default_ctlog_list_file() - how does that sound? The original author was seemingly aiming for consistency first and foremost.

File: apps/apps.c (hunk 2 above)
Comment (RJPercival) [inline apps/apps.c]: Done in apps/apps.h.

**Comment by ekasper** [comment]
1) Suppressing errors in s_client.c but not in general is the right solution. Please update the doc; not urgent today, though.
2) A dedicated `set_default` method is better than polluting public headers with the paths, imho. Modelling old APIs after new ones is not always wise but in this case, it seems to make sense.

**Comment by ekasper** [comment]
Ah - I see you already updated the docs :) +1

**Comment by richsalz** [comment]
@RJPercival can you do a quick rebase against master?  I'm getting several conflicts and don't want to screw it up :)

**Comment by RJPercival** [comment]
Rebased.

**Comment by richsalz** [comment]
thank you sir.  pushed.
if you're still looking for something to do :) how about a policy that implements the Chrome EV/SCT callback?

**Comment by RJPercival** [comment]
I'm working on "-precert" and "-sctsfile" flag for "openssl ca" right now, so that OpenSSL can generate pre-certificates and embed SCTs in certificates (the API supports this, so there's just a little code required in apps/ca.c to expose it). I should have time to add such a policy as well though. That's just a callback that insists on at least two valid SCTs, one from a Google log and one from any other log, right?

**Comment by richsalz** [comment]
Those flags are more important; the callback (you have the semantics right) can always be shipped in demos.  Hmm, in fact that makes the demo directory worthwhile ☺

**Pull request closed without merging**